### DISCUSSION TITLE
Multiplexing 7-Segment Display

### DISCUSSION THREAD

**Discussion by tblazey** [micropython/micropython#9453] [Using MicroPython]
Hi,

I am trying to use a Raspberry Pi Pico and a [CD4094BE shift register ](https://www.digikey.com/en/products/detail/texas-instruments/CD4094BE/67331) to multiplex a common cathode [25 mm 7-segment display](https://www.newark.com/forge/fn1-1001w4sm00bw/display-seven-segment-25mm-white/dp/06AC5994?CMP=e-email-sys-invoice-NA-Product). I am using the CD4094BE as it allows for greater voltages then the Pico can provide. Also eventually I want to multiplex several 7-segment displays instead of multiplexing individual LEDs on a single display. I have attached a diagram of my wiring setup and I am using MicroPython v1.19.1 ( installed via Thonny).

Below is the code I came up with to display an '8' on the 7-segment display by quickly turning on/off individual LEDs. 

```py
from` machine import Pin, SPI
import time

# Pin numbers
latch_gp = 13
sck_gp = 14
tx_gp = 15
oe_gp = 16

# Assign pins
latch_pin = Pin(latch_gp, Pin.OUT)
oe_pin = Pin(oe_gp, Pin.OUT, value=1)

# Construct SPI interface
spi = SPI(1, baudrate=16000, sck=Pin(sck_gp), mosi=Pin(tx_gp))

# Function for communicating with shift register
def shift_bits(byte_list, spi, latch):
    
    #Send data to register.
    spi.write(bytearray(byte_list))
              
    #Send data to outputs
    latch.value(1)
    latch.value(0)

# Show "8" for 15 seconds
start_time = time.ticks_us()
while time.ticks_diff(time.ticks_us(), start_time) < 15E6:
    
    # Turn on each led in turn
    for led in [1, 2, 4, 16, 32, 64, 128]:
        shift_bits([led], spi, latch_pin)
        time.sleep_us(500)

# Turn off all outputs
shift_bits([0], spi, latch_pin)
```
While this code does turn on the correct LEDs, it will occasionally flicker. I think I am refreshing the display fast enough, and the flicker persists even if I take the refresh time down to 1 microsecond. I've also tried adding a time constraint within the for loop, but that didn't help:

```py
    for led in [1, 2, 4, 16, 32, 64, 128]:
        led_start = time.ticks_us()
        shift_bits([led], spi, latch_pin)
        time.sleep_us(500 - time.ticks_diff(time.ticks_us(), led_start))
```

I also experimented with asyncio as suggested in this [thread](https://forum.micropython.org/viewtopic.php?t=12274&p=66651), and that didn't fix the issue either. I'm still new to electronics, so apologies if I am making a silly error. 

Thanks so much for any help!

-Tyler

![multiplex_help_bb](https://user-images.githubusercontent.com/810152/192881669-a1c74575-bfd6-40df-8b2f-ada45d52a95e.jpg)

**Comment by mcauser** [comment]
I did something similar with the 74HC595 shift register, using bit bang and SPI:
https://github.com/mcauser/micropython-74hc595
Problem with shift registers persistence of vision exploit is the subtle flicker when the controller is doing something else.

For flicker-free, I prefer using an external controller, such as the TM1637:
https://github.com/mcauser/micropython-tm1637
You tell it once what you want it to display, then your mcu is free to do other things.

If you want more than 4x 7-segments, there's other TM chips with more IOs:
https://github.com/mcauser/micropython-tm1638
https://github.com/mcauser/micropython-tm1640

**Reply by tblazey** [reply]
Thanks! I will take a look at the external controllers that you suggested.

-Tyler

**Comment by OlivierLenoir** [comment]
You can find a driver for SN74HCS264, 8-Bit Parallel-Out Serial Shift Registers With Schmitt-Trigger Inputs and Inverted Outputs
,  https://gitlab.com/olivierlenoir/MicroPython-SN74HCS264
